src/adapter/src/rbac.rs

```diff
@@ -51,70 +66,115 @@ impl std::fmt::Display for Attribute {
     }
 }
 
-/// Errors that can occur due to an unauthorized action.
 #[derive(Debug)]
-pub struct UnauthorizedError {
-    /// The action that was forbidden.
-    action: String,
-    kind: UnauthorizedErrorKind,
+pub struct Ownership(ObjectId);
+
+impl Ownership {
+    /// Reports whether a role has ownership over an object.
+    fn check_role(&self, role_id: &RoleId, catalog: &impl SessionCatalog) -> bool {
+        match self.0 {
+            ObjectId::Cluster(id) => catalog.get_cluster(id).owner_id() == *role_id,
+            ObjectId::ClusterReplica((cluster_id, replica_id)) => {
+                catalog
+                    .get_cluster_replica(cluster_id, replica_id)
+                    .owner_id()
+                    == *role_id
+            }
+            ObjectId::Database(id) => catalog.get_database(&id).owner_id() == *role_id,
+            ObjectId::Schema((database_id, schema_id)) => {
+                catalog
+                    .get_schema(
+                        &ResolvedDatabaseSpecifier::Id(database_id),
+                        &SchemaSpecifier::Id(schema_id),
+                    )
+                    .owner_id()
+                    == *role_id
+            }
+            ObjectId::Item(id) => catalog.get_item(&id).owner_id() == *role_id,
+        }
+    }
+
+    /// Reports whether any role has ownership over an object.
+    fn check_roles(&self, role_ids: &BTreeSet<RoleId>, catalog: &impl SessionCatalog) -> bool {
+        role_ids
+            .iter()
+            .any(|role_id| self.check_role(role_id, catalog))
+    }
 }
 
+/// Errors that can occur due to an unauthorized action.
 #[derive(Debug)]
-enum UnauthorizedErrorKind {
+pub enum UnauthorizedError {
```

Comment:
nit: it could be nice to use `thiserror::Error` here, which would remove the need to implement Display manually

Reply:
I made the change, but I'm not in love with including non-trivial display logic in an attribute. So I may revert it unless there's a better way.